### Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

### Import Prods and Ords

In [3]:
df_prods = pd.read_csv(r'C:\Users\PC\Documents\Instacart Basket Analysis\02 Data\Original Data\4.3_orders_products\products.csv', index_col = False)

In [5]:
df_ords = pd.read_csv(r'C:\Users\PC\Documents\Instacart Basket Analysis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

### Dataframe creation

In [7]:
df_test = pd.DataFrame()

### Create a mixed type column

In [13]:
df_test['mix'] = ['a', 'b', 1, True]

In [15]:
df_test.head()

mix
0     a
1     b
2     1
3  True

### Function for checking any mixed types

In [18]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)

mix


### Changing Data Type

In [21]:
df_test['mix'] = df_test['mix'].astype('str')

### Finding Missing Values

In [24]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

### Subsetting (Filtering)

In [27]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [29]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Checking rows

In [32]:
df_prods.shape

(49693, 5)

### Filtering out missing values

In [35]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

### Checking rows of new dataframe

In [38]:
df_prods_clean.shape

(49677, 5)

### Duplicates

In [41]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [43]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Removing Duplicates

In [46]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [48]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

### ### STEP 2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [51]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  Placing_orders_time  days_since_prior_order  
count        3.421083e+06         3.421083e+06            3.214874e+06  
mean         2.776219e+00         1.345202e+01            1.111484e+01  
std          2.046829e+00         4.226088e+00            9.206737e+00  
min          0.000000e+00         0.000000e+00            0.000000e+00  
25%          1.000000e+00         1.000000e+01            4.000000e+00  
50%          3.000000e+00         1.300000e+01            7.000000e+00  
75%          5.000000e+00         1.600000e+01            1.500000e+01  
max          6.000000e+00         2.300000e+01            3.000000e+01

### ### The following values should be further investigated: orders_day_of_week: Min 0, Placing_order_time: Min 0

### STEP 3 Check for mixed-type data in your df_ords dataframe.

In [55]:
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].map(type)).any(axis=1)
    if len(df_ords[weird]) > 0:
        print(col)

### No mixed - type data were finded.

### STEP 5 Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [59]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
Placing_orders_time            0
days_since_prior_order    206209
dtype: int64

### days_sice_prior_order 206209, it could be due to reasons: 
1.- It was the first order of the customer (There is no previous order).
2.- When the data was downloaded, the customer could just have ordered on the same day previous the download.

### STEP 6 Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice

In [64]:
df_ords.shape

(3421083, 7)

### 206209 missing vales represents more than the 5% of the entire data (3421083), it is adviced to do nothing.

### STEP 7 Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [68]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [70]:
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, Placing_orders_time, days_since_prior_order]
Index: []

### There are no duplicated values in the orders data.

### Step 9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [76]:
df_ords.to_csv(r'C:\Users\PC\Documents\Instacart Basket Analysis\02 Data\Prepared Data\orders_cleaned1.csv')

In [78]:
df_prods_clean_no_dups.to_csv(r'C:\Users\PC\Documents\Instacart Basket Analysis\02 Data\Prepared Data\products_cleaned1.csv')